[Lin Chieh] iP

euph00 opened the pull request:
# CS2103T iP
## Chatbot by Lin Chieh - Duke
---
> "Your enemies cant predict what you're doing if you don't know what you're doing yourself" – Ludwig van Beethoven, in The Communist Manifesto
- Not the best code quality
- OOP designs *may* not always make sense
- Please do suggest improvements or corrections 🥺 
- Is trying his best 😔 

To test the app:
1. Download the latest release [here](https://github.com/euph00/ip)
2. I'm not sure what's next, might be replacing the text ui with javafx gui very soon

if you download the source code instead this is what the main method looks like for now
`public static void main(String[] args) {
        new Duke("src/data/duke.txt").run();
    }`

Week 4 tasks:
- [ ] Checkstyle
- [ ] GUI
- [x] **PROMOTE HOSHIMACHI SUISEI'S NEW ALBUM, [SPECTER](https://cover.lnk.to/Suisei2ndAlbum), WHICH IS NOW ON MAJOR STREAMING SITES**

Commit:
Add Gradle support

Commit:
Completed Level-1, skeletal version of chatbot

Commit:
Level-2 completed, add and list items

Commit:
Level-3 completed, mark and unmark functionality

Commit:
Level-4 Added Todo, Deadline, Event task types

Commit:
automated UI testing

Commit:
implemented exception handling, checking for common exceptions

Commit:
implemented delete functionality

Commit:
Implement saving of task list to the hard drive

Commit:
Implement partially the DateTime feature for Event and Deadline

Commit:
Implements recognising date and time for Event and Deadline as localdatetime objects, with custom output display

Commit:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Duke.java

Commit:
Merge Level 7 and Level 8 branches, allowing for functionality of recognising localdatetime and saving of state in master branch

Commit:
Add duke.txt to gitignore

Commit:
Refactor the code to use OOP for Ui, Storage, TaskList and Parser classes, distributing the code from driver class

Commit:
Package java class files under appropriate names

Commit:
Merge remote-tracking branch 'origin/add-gradle-support'

Commit:
Implement unit testing with Junit for a few methods

Commit:
Not affect anything substantial, only to include the JAR release

Commit:
Include javadocs for most non private classes and methods

To include javadocs for Task related classes at later date

Commit:
Implement the find feature

Users can now input the find command followed by any number of keywords to search for Tasks that have names containing the search target keywords

Commit:
Modify code to comply with coding standard, minimal changes made, will revisit with checkstyle

Commit:
Merge branch 'branch-A-JavaDoc'

Commit:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/TaskList.java
#	src/main/java/duke/Ui.java

Commit:
Partially implement the GUI

This commit does not compile. This is just a temporary save point.

Commit:
Implement working barebones version of GUI

Automated regression and unit testing cases need to be completely revamped. They are currently meaningless.

Commit:
Implement fully the transition to GUI with relevant FXML files in place

Commit:
Update Javadocs to be consistent with GUI implementation

euph00 committed:
Merge pull request #1 from euph00/branch-Level-10

Branch level 10

sanjevi13 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,140 @@
+    }
+
+    /**
+     * Returns a <code>Message</code> that represents the response of the chatbot when prompted by an input by user.
```
I think for the description try not to include syntax like the code syntax. I believe this part of the JavaDocs is just supposed to tell us in general what the function does, not what it returns. Otherwise, well done! I appreciate the fact that you put in a great amount of effort into your JavaDocs

sanjevi13 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,140 @@
+     * @param msg A <code>Message</code> object representing the user's request
+     * @return A <code>Message</code> object containing the response of the bot to the user's request
+     */
+    public Message respondToMessage(Message msg) {
```
Good job on splitting up the input commands, the method is relatively easy to read! Is there a way that you can be able to do this parsing in your Parser class instead?

sanjevi13 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,296 @@
+     * and adds the contents into the current session.
+     * @param tokens <code>String[]</code> of tokens as read from the data file by <code>Parser</code>.
+     */
+    void parseEventFromFile(String[] tokens) {
```
Nice abstraction!

sanjevi13 commented on src/main/java/duke/ui/DialogBox.java:
```diff
@@ -0,0 +1,61 @@
+import javafx.fxml.FXMLLoader;
+import javafx.geometry.Pos;
+import javafx.scene.Node;
+import javafx.scene.control.Label;
```
Maybe can order the import statements in alphabetical order? Although I'm not too sure of the convention, as long as your ordering of imports is consistent in all files, it should be fine!

sanjevi13 reviewed (commented):
Good job! I think the app in general is very well done and I appreciate the fact that you took the effort to make separate packages for different code components. The changes I suggested are minor, and they include slight changes for the description part of the JavaDocs for classes and methods. Overall, most of your code is easy to read and understand. Otherwise, keep up the good work!

Douglch commented on src/main/java/duke/TaskList.java:
```diff
@@ -43,26 +44,20 @@ void parseEventFromFile(String[] tokens) {
      */
-    void addToDo(String[] tokens, Ui ui) throws DukeException {
+    ToDo addToDo(String[] tokens) throws DukeException {
         StringBuilder sb = new StringBuilder();
```
Perhaps you could instantiate one StringBuilder instance class wide, then calling sb.setLength(0) to clear the sb at the start of a method, so that you don't have to create multiple sb instances.

Douglch commented on src/main/java/duke/Duke.java:
```diff
@@ -1,91 +1,140 @@
+        String input = msg.getMessage();
+        String[] tokens = parser.parseUserInput(input);
+
+        if (tokens.length == 1 && tokens[0].equals("bye")) {
```
I think there is potential to reduce the clutter in the Duke class, perhaps you could refactor them with a class for each command and handle inputs in the parser (which would call the commands separately)?

Douglch commented on src/main/java/duke/TaskList.java:
```diff
@@ -149,85 +147,83 @@ void addEvent(String[] tokens, Ui ui) throws DukeException {
-    void deleteItem(String[] tokens, Ui ui) throws DukeException {
+    Task deleteItem(String[] tokens, UiController ui) throws DukeException {
         if (tokens.length != 2) {
             throw new DukeException("please specify delete command as delete [list index]");
```
Good to see that you've made many exceptions tailoring to different types of user inputs👍

Douglch commented on src/main/java/duke/TaskList.java:
```diff
@@ -243,7 +239,42 @@ void findItemInList(String[] tokens, Ui ui) throws DukeException {
+        sb.append("Here are the tasks in your list:\n");
+        for (int i = 0; i < tasks.size(); i++) {
+            sb
+                    .append(i+1)
```
Think there should be whitespaces in between the operands and the operator

Douglch reviewed (commented):
The code looks pretty clean to me, I think except for some parts which you can refactor a little more and the point on the Javadocs brought up by the other reviewer, its all gucci 🥇

GlennOngJunJie commented on src/main/java/Duke.java:
```diff
@@ -1,273 +1,65 @@
-                duke.exit();
+            String[] tokens = parser.parseUserInput();
+            if (tokens.length == 1 && tokens[0].equals("bye")) {
+                exit();
```
Love the idea of using the parser to break down the sentences! :) but could also consider using a switch to improve readability, but this works well too! :)

GlennOngJunJie commented on src/main/java/duke/TaskList.java:
```diff
@@ -1,3 +1,11 @@
+package duke;
+
+import duke.exceptions.DukeException;
+import duke.task.Deadline;
+import duke.task.Event;
+import duke.task.Task;
+import duke.task.ToDo;
```
It is wonderful that you imported the packages one by one and not just import everything wholesale!

GlennOngJunJie commented on src/main/java/duke/exceptions/DukeException.java:
```diff
@@ -1,3 +1,5 @@
+package duke.exceptions;
```
Great that you created more than one package to improve readability

GlennOngJunJie reviewed (commented):
Overall very good job! The code is very readable and easy to follow, gave some insights for me to improve my code too

Commit:
Introduce the use of java assertions to verify invariants

Several invariants, such as the existence of key objects and the
correctness of arguments when calling certain functions are to be
verified before these functions can be safely called. Otherwise,
there is a chance of undefined or unexpected behavior. This commit
uses java assertions to verify some of these invariants.

Commit:
Update code quality, reducing reliance on long chains of if else
by replacing it with a switch case, and updating junit test cases

euph00 committed:
Merge pull request #2 from euph00/heads/A-CodeQuality

Update code quality, reducing reliance on long chains of if else by replacing it with a switch case, and updating junit test cases

Commit:
Implement the tagging feature and command, allowing users to specify
a list index and add a tag to the event

euph00 committed:
Merge pull request #3 from euph00/branch-C-Tagging

Implement the tagging feature and command, allowing users to specify

Commit:
Improve code quality by abstracting sections of
Duke::respondToMessage that was deemed too long, and rectify the
fact that I forgot to branch for adding assertions

euph00 committed:
Merge pull request #5 from euph00/branch-A-Assertions

Improve code quality by abstracting sections of Duke::respondToMessage that was deemed too long, and rectify the fact that I forgot to branch for adding assertions

Commit:
Improve code quality and resolve issues raised by grader script

euph00 committed:
Add files via upload

euph00 committed:
Update README.md

euph00 committed:
Update the README with a user guide for Duke

Commit:
Update build.gradle for shadowJar to work

Commit:
Merge branch 'master' of https://github.com/euph00/ip

euph00 committed:
Delete Ui.png

euph00 committed:
Add files via upload

Commit:
Add a window title to duke

Commit:
Merge branch 'master' of https://github.com/euph00/ip

Commit:
Refactor code slightly for java 11
